src/Symfony/Component/Form/Util/ParamFilesMerger.php

```diff
@@ -0,0 +1,184 @@
+
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+/**
+ * @author Priyadi Iman Nurcahyo <priyadi@rekalogika.com>
+ */
+class ParamFilesMerger
+{
+    private array $path;
+    private array $params;
+    private array $files;
+
+    /**
+     * @param array $path   The path to the current element, empty means the root
+     * @param array $params The parameters
+     * @param array $files  The files
+     */
+    public function __construct(array $path, array $params, array $files)
+    {
+        $this->path = $path;
+        $this->params = $params;
+        $this->files = $files;
+    }
+
+    public function getResult()
+    {
+        $paramsValue = $this->getParamsValue();
+        $filesValue = $this->getFilesValue();
+
+        if (null === $paramsValue) {
+            if (null === $filesValue) {
+                return null;
+            } elseif (self::isFileUpload($filesValue)) {
+                return $filesValue;
+            } elseif (\is_array($filesValue)) {
+                return $filesValue;
+            } else { // $filesValue has a non-array value
+                return $filesValue;
+            }
+        } elseif (\is_array($paramsValue)) {
+            if (null === $filesValue) {
+                return $paramsValue;
+            } elseif (self::isFileUpload($filesValue)) {
+                return $filesValue; // if the array is a file upload field, it has the precedence
+            } elseif (\is_array($filesValue)) {
+                return $this->getResultArray($paramsValue, $filesValue);
+            } else { // $filesValue has a non-array value
+                return $paramsValue; // params has the precedence
+            }
+        } else { // $paramsValue has a non-array value
+            if (null === $filesValue) {
+                return $paramsValue;
+            } elseif (self::isFileUpload($filesValue)) {
+                return $filesValue; // if the array is a file upload field, it has the precedence
+            } elseif (\is_array($filesValue)) {
+                return $paramsValue; // params has the precedence
+            } else { // $filesValue has a non-array value
+                return $paramsValue; // params has the precedence
+            }
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private static function isFileUpload($value)
+    {
+        if ($value instanceof UploadedFile) {
+            return true;
+        }
+
+        if (!\is_array($value)) {
+            return false;
+        }
+
+        $keys = array_keys($value);
+        sort($keys);
+
+        return $keys === ['error', 'name', 'size', 'tmp_name', 'type'];
+    }
+
+    /**
+     * @param array $array
+     *
+     * @return bool
+     */
+    private static function doesNotContainArrayOrFileUpload($array)
+    {
+        foreach ($array as $value) {
+            if (
+                \is_array($value)
+                && !self::isFileUpload($value)
+            ) {
```

Comment:
I think this condition is simple enough to be in one line

Reply:
done